### PR TITLE
Add support for non-square flats; improve texture/sprite caching

### PR DESCRIPTION
**Pull request by dashodanger**
This removes the requirement for equal height and width when loading a flat in the Texture Manager; as well as updating the drawing routines to account for flats with dimensions other than 64 x 64.

Also, sprite and texture caching in the Texture Manager has been cleaned up; the images will now be directly transformed from RGBA->tallpost patch instead of RGBA->regular post patch->tallpost patch. I've also cleaned up some now-unneeded functions and structs related to conversion of traditional Doom patches.

### CODE REVIEW

File: source/engine/client/r_draw.cpp [hunk 1]
```diff
@@ -685,17 +685,27 @@ static forceinline void R_DrawLevelSpanGeneric(PIXEL_T *dest, const drawspan_t &
...
         return;
 
-    dsfixed_t       xfrac = drawspan.xfrac;
-    dsfixed_t       yfrac = drawspan.yfrac;
-    const dsfixed_t xstep = drawspan.xstep;
-    const dsfixed_t ystep = drawspan.ystep;
+    // These are intentionally swapped - Dasho
+    const uint32_t width_bits = dspan.texture_height_bits;
+	const uint32_t height_bits = dspan.texture_width_bits;
+
```
Comment (pbdot) [inline source/engine/client/r_draw.cpp:691]: This swap is a bit awkward, I see the comment, though maybe we can sort the names? xbits/ybits? ubits/vbits?  Not sure what the logic is..

File: source/engine/client/r_draw.cpp [hunk 2]
```diff
@@ -744,6 +754,15 @@ static forceinline void R_DrawSlopedSpanGeneric(PIXEL_T *dest, const drawspan_t
 
     int32_t ltindex = 0;
 
+    // These are intentionally swapped - Dasho
+    const uint32_t width_bits = dspan.texture_height_bits;
+	const uint32_t height_bits = dspan.texture_width_bits;
```
Comment (pbdot) [inline source/engine/client/r_draw.cpp:759]: Same thing on this swap

File: source/engine/client/r_drawt_sse2.cpp [hunk 3]
```diff
@@ -193,6 +195,15 @@ void R_DrawSlopeSpanD_SSE2(void)
 
     int32_t ltindex = 0; // index into the lighting table
 
+    // These are intentionally swapped - Dasho
+    const uint32_t width_bits = dspan.texture_height_bits;
+	const uint32_t height_bits = dspan.texture_width_bits;
+
```
Comment (pbdot) [inline source/engine/client/r_drawt_sse2.cpp:201]: Here too :)

File: source/engine/common/r_draw.h [hunk 4]
```diff
@@ -72,6 +72,8 @@ typedef struct
     int32_t x1;
     int32_t x2;
 
+    uint32_t texture_width_bits;
+	uint32_t texture_height_bits;
     dsfixed_t xfrac;
```
Comment (pbdot) [inline source/engine/common/r_draw.h:77]: There is some funky whitespace, we should run a clang format pass soon, which will also catch anything introduced with header ordering.  I wouldn't worry about it for this PR though

**Review by pbdot — COMMENTED** [review]
Looks great, just a couple comments.

File: source/engine/client/r_draw.cpp (hunk 1 above)
Comment (dashodanger) [reply source/engine/client/r_draw.cpp:690]: Updated the naming to have 'ubits' as the texture_height_bits value and 'vbits' as the texture_width_bits value

File: source/engine/client/r_draw.cpp (hunk 2 above)
Comment (dashodanger) [reply source/engine/client/r_draw.cpp]: Fixed (see above)

File: source/engine/client/r_drawt_sse2.cpp (hunk 3 above)
Comment (dashodanger) [reply source/engine/client/r_drawt_sse2.cpp:199]: Fixed (see above the above)

File: source/engine/common/r_draw.h (hunk 4 above)
Comment (dashodanger) [reply source/engine/common/r_draw.h:77]: In my local VSCode, the whitespacing here isn't present for me; not sure how that's being introduced honestly

Comment (pbdot) [reply source/engine/common/r_draw.h:77]: Line 76 uses a tab instead of spaces, and your editor is probably set to 4 space tabs

**Pull request merged**